refactor(ivy): move classMap interpolation logic internally

crisbeto opened the pull request:
Adds the new `classMapInterpolate1` through `classMapInterpolate8` instructions which handle interpolations inside the `class` attribute and moves the interpolation logic internally. This allows us to remove the `interpolationX` instructions in a follow-up PR.

kara commented on packages/compiler/src/render3/view/styling_builder.ts:
```diff
@@ -377,22 +379,32 @@ export class StylingBuilder {
     if (mapValue instanceof Interpolation) {
+      if (!isClassBased) {
+        // We only support interpolations inside class bindings.
+        error('Unexpected interpolation');
```
Can we add some more info in the error? Like

```suggestion
    error('Unexpected interpolation. Interpolated values are not supported in the `style` property.');
```

kara commented on packages/compiler/src/render3/view/styling_builder.ts:
```diff
@@ -377,22 +379,32 @@ export class StylingBuilder {
+      }
+
       totalBindingSlotsRequired += mapValue.expressions.length;
+      reference = isClassBased ? getClassMapInterpolationExpression(mapValue) : R3.styleMap;
```
The fact that we fall back to `R3.styleMap` here seems strange since we are throwing an error on line 386 if `!isClassBased`.  Seems like we can assume at this point that it's class-based, no?

```suggestion
      reference = getClassMapInterpolationExpression(mapValue);
```

kara reviewed (changes requested):
A few small things below. 

Seems like we also want to remove `NO_CHANGE` from the `classMap` signature.

crisbeto committed:
refactor(ivy): move classMap interpolation logic internally

Adds the new `classMapInterpolate1` through `classMapInterpolate8` instructions which handle interpolations inside the `class` attribute and moves the interpolation logic internally. This allows us to remove the `interpolationX` instructions in a follow-up PR.

These changes also add an error if an interpolation is encountered inside a `style` tag (e.g. `style="width: {{value}}"`). Up until now this would actually generate valid instructions, because `styleMap` goes through the same code path as `classMap` which does support interpolation. At runtime, however, `styleMap` would set invalid styles that look like `<div style="0:w;1:i;2:d;3:t;4:h;5::;7:1;">`. In `ViewEngine` interpolations inside `style` weren't supported either, however there we'd output invalid styles like `<div style="unsafe">`, even if the content was trusted.

crisbeto commented:
Reworked based on the feedback @kara.

kara reviewed (approved):
LGTM

kara commented:
[VE presubmit](http://test/OCL:255432399:BASE:255432442:1561657442121:db62704c)

[Ivy presubmit](http://test/OCL:255432399:BASE:255432703:1561657500404:5ac56272)

kara commented:
Some G3 failures from the presubmit with `Unexpected interpolation` error. Need to investigate.

AndrewKushnir commented:
@crisbeto I performed an investigation and the problem was caused by the presence of interpolation in style attrs, here is an example template:

```
  <div style.background-image="url({{myImageUrl}})"></div>
```

VE outputs the following:

```
  <div style="background-image:url(http://test.com/img.png);"></div>
```

i.e. it looks like interpolations are supported in style attrs in VE.

crisbeto commented:
Thank you for looking into it @AndrewKushnir! I've removed the error and added interpolation instructions for `styleMap` as well. Can you take another look @kara? All the next changes are in the last fixup commit. Note that I haven't added any runtime tests, because they'll fail. I've sent a couple of test cases to Matias to try against the new styling implementation.

crisbeto commented:
Updated to add instructions for `styleProp` and remove the ones for `styleMap`.

crisbeto committed:
fixup! refactor(ivy): move classMap interpolation logic internally

kara reviewed (approved):
LGTM

kara commented:
[New VE presubmit](http://test/OCL:256095889:BASE:256095902:1562043169233:627f1d3c)
[New Ivy presubmit](https://test.corp.google.com/ui#id=OCL:256095889:BASE:256096107:1562043333733:b7bc0e75)

kara commented:
merge-assistance: global
